add pairwise! for Covariance types

markmbaum opened the pull request:
This adds the `pairwise!` function for `Covariance` types. I also added a docstring for the variogram's `pairwise!` method.

closes JuliaEarth/GeoStats.jl#405

I did also notice that the existing implementation allocates a lot of memory even though it should just be filling a destination matrix. For example:
```julia
julia> grid = CartesianGrid((2, 3))
2×3 CartesianGrid{2,Float64}
  minimum: Point(0.0, 0.0)
  maximum: Point(2.0, 3.0)
  spacing: (1.0, 1.0)

julia> A = zeros(6, 6)
6×6 Matrix{Float64}:
 0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0

julia> γ = GaussianVariogram()
GaussianVariogram
├─ sill: 1.0
├─ nugget: 0.0
├─ range: 1.0
└─ distance: Euclidean

julia> @benchmark GeoStatsFunctions.pairwise!($A, $γ, $grid)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  429.125 μs …   1.775 ms  ┊ GC (min … max): 0.00% … 72.74%
 Time  (median):     438.125 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   462.660 μs ± 165.411 μs  ┊ GC (mean ± σ):  5.08% ± 10.09%

  █▃                                                            ▁
  ██▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇█▆▄▃▃▁▁▁▁▁▁▅ █
  429 μs        Histogram: log(frequency) by time       1.69 ms <

 Memory estimate: 948.28 KiB, allocs estimate: 7707.
```
I'm not exactly sure why and I didn't want to mess with the original implementation.

Commit:
add pairwise! for Covariance types

juliohm commented on src/covariance.jl:
```diff
@@ -18,15 +18,30 @@ Evaluate the covariance at objects `x₁` and `x₁`.
+Evaluates covariance `cov` between all elements of `domain₁` and `domain₂` in-place, filling the matrix `Γ`."""
+function pairwise!(Γ, cov::Covariance, args...)
+  pairwise!(Γ, cov.γ, args...)
+  Γ .*= -one(eltype(Γ))
+  Γ .+= sill(cov.γ)
+  Γ
```
Let's use a simple for loop here to modify the elements of the matrix instead of dot syntax.

juliohm reviewed (changes requested):
The test with CartesianGrid allocates because it computes covariance between the elements of the domain, in this case quadrangles. In order to compute the covariance between two quadrangles, the variogram samples points inside the quadrangles based on the range, and performs numerical integration.

The variogram and covariance don't require allocation when they are evaluated with points. For example, if you construct PointSet(centroid.(grid)) as your domain, that will be much faster and shouldn't allocate.

Please add tests here as well for coverage purposes.

Commit:
add pairwise! covariance test and update function

markmbaum commented:
Should be all set. I added a test that will cover the new `pairwise!` method and changed the implementation to a loop.

This is still a little off topic, but I profiled `pairwise!` with a `PointSet` and still found some memory allocation.
```julia
julia> Γ = Matrix{Float64}(undef, 3, 3)
3×3 Matrix{Float64}:
 1.0e-323  4.4e-323  1.2e-322
 2.5e-323  9.4e-323  1.5e-322
 4.0e-323  1.1e-322  0.0

julia> c = GaussianCovariance()
GaussianCovariance
├─ sill: 1.0
├─ nugget: 0.0
├─ range: 1.0
└─ distance: Euclidean

julia> 𝒟 = PointSet(Matrix(1.0I, 3, 3))
3 PointSet{3,Float64}
├─ Point(1.0, 0.0, 0.0)
├─ Point(0.0, 1.0, 0.0)
└─ Point(0.0, 0.0, 1.0)

julia> ProfileView.@profview begin
         for i ∈ 1:1_000_000
           GeoStatsFunctions.pairwise!(Γ, c, 𝒟)
         end
       end
```

Which looks like it's coming from the `variosample` calls in the methods for `Variogram` types.

<img width="793" alt="Screenshot 2024-03-24 at 6 55 50 PM" src="https://github.com/JuliaEarth/GeoStatsFunctions.jl/assets/18745581/070cfe9f-53cb-46b4-9cfb-268a47469303">

juliohm commented on src/covariance.jl:
```diff
@@ -18,15 +18,31 @@ Evaluate the covariance at objects `x₁` and `x₁`.
 pairwise(cov::Covariance, args...) = sill(cov.γ) .- pairwise(cov.γ, args...)
 
+"""
+    pairwise!(Γ, cov, domain)
```
We denote the covariance matrix by C or \Sigma. \Gamma is reserved for the variogram matrix.

juliohm commented:
Left a small comment regarding notation. Can you please replace \Gamma --> C or \Sigma in the covariance methods?

Also, it would be nice to investigate this allocation with Point in another PR. We probably need an extra method to avoid sampling the trivial geometry (i.e., point).

juliohm committed:
Minor adjustments

juliohm commented on src/variogram.jl:
```diff
@@ -185,6 +190,11 @@ function pairwise(γ::Variogram, domain₁, domain₂)
 end
 
+"""
+    pairwise!(Γ, γ, domain)
+
+Evaluates covariance `γ` between all elements of `domain₁` and `domain₂` in-place, filling the matrix `Γ`.
```
```suggestion
    pairwise!(Γ, γ, domain₁, domain₂)

Evaluates covariance `γ` between all elements of `domain₁` and `domain₂` in-place, filling the matrix `Γ`.
```

juliohm committed:
Update src/variogram.jl

juliohm commented:
@markmbaum this PR has conflicts with the other PR that got merged.

juliohm committed:
Merge branch 'main' into covariance-pairwise
